src/utils/equipment.ts

```diff
@@ -0,0 +1,83 @@
+import Equipment from "../resources/Equipment";
```

Comment:
Suggested refactoring.
This uses the old `Category: {name: string;}` definition.

```js
type ItemTest = (item: Equipment) => boolean;
type StringBoolMap = { [k: string]: boolean };

// returns keys ready to compare
const extractStrings = (mappedStrings: StringBoolMap): string[] =>
  Object.keys(mappedStrings).reduce(
    (strings, key) =>
      mappedStrings[key] ? [...strings, key.toLowerCase().trim()] : strings,
    [] as string[]
  );

// if search is empty, defaults to TRUE
const makeSearchStringTest = (search: string): ItemTest => {
  const tokens = search.split(/\s+/).map((t) => t.toLowerCase().trim());
  return ({ description, category, tags }: Equipment): boolean =>
    tokens[0]
      ? tokens.some(
          (token) =>
            description.includes(token) ||
            category.name.includes(token) ||
            tags.some(({ name }) => name.includes(token))
        )
      : true;
};

// if filters are empty, defaults to TRUE
const makeFiltersTest = (filters: StringBoolMap): ItemTest => {
  const tags = extractStrings(filters);
  return (e: Equipment): boolean =>
    tags.length
      ? tags.every((tag) =>
          e.tags.some(({ name }) => name.toLowerCase().includes(tag))
        )
      : true;
};

// specialized "clean up" processor for Equipment, could be a class method
const lowerCaseAndTrimEquipment = (e: Equipment): Equipment => ({
    ...e,
    category: { ...e.category, name: e.category.name.toLowerCase().trim() },
    description: e.description.toLowerCase().trim(),
    tags: e.tags.map((t) => ({ ...t, name: t.name.toLowerCase().trim() })),
  });

// reduce all tests into a single predicate fn for Array.prototype.filter
const makeItemTest = (
  searchString: string,
  filters: StringBoolMap
): ItemTest => {
  const stringTest = makeSearchStringTest(searchString);
  const filterTest = makeFiltersTest(filters);
  return (item: Equipment): boolean => {
    const itemUnderTest = lowerCaseAndTrimEquipment(item);
    return stringTest(itemUnderTest) && filterTest(itemUnderTest);
  };
};

// the exported wrapper
export const filterEquipment = (
  equipment: Equipment[],
  searchString: string,
  filters: StringBoolMap
): Equipment[]  => {
  const test = makeItemTest(searchString, filters);
  return equipment.filter(test);
};
```

Reply:
The logic to the above suggested refactoring is that we have a clean pipeline of
* tests are setup once (including any cleaning of test inputs)
* item is cleaned once
* each item goes through the test pipeline once

The emphasis on "once" is not for performance (although surely that doesn't hurt).
The benefit is that we can localize that step to one block of code.
We can read & reason with more confidence.